Clarify definition of phylogenetic tree and adjust TreeTopology, Tree, TreeSpace, and BHVMetric accordingly

abertics opened the pull request:
<!--
Thank you for opening this pull request!
-->

## Checklist

- [ ] My pull request has a clear and explanatory title.
- [ ] If necessary, my code is [vectorized](https://www.geeksforgeeks.org/vectorization-in-python/).
- [ ] I added appropriate unit tests.
- [ ] I made sure the code passes all unit tests. (refer to comment below)
- [ ] My PR follows [PEP8](https://peps.python.org/pep-0008/) guidelines. (refer to comment below)
- [ ] My PR follows [geomstats coding style](https://github.com/geomstats/geomstats/blob/main/docs/contributing.rst#coding-style-guidelines) and API.
- [ ] My code is properly documented and I made sure the documentation renders properly. ([Link](https://github.com/geomstats/geomstats/blob/main/docs/contributing.rst#documentation))
- [ ] I linked to issues and PRs that are relevant to this PR.


<!-- For checking consistency of entire codebase
First, run the tests related to your changes. For example, if you changed something in geomstats/spd_matrices_space.py:
$ pytest tests/tests_geomstats/test_spd_matrices.py

and then run the tests of the whole codebase to check that your feature is not breaking any of them:
$ pytest tests/

This way, further modifications on the code base are guaranteed to be consistent with the desired behavior. Merging your PR should not break any test in any backend (numpy, autograd or pytorch)."

For testing in alternative backends such as `numpy`, `pytorch`, `autograd`, set the environment variable using:
$ export GEOMSTATS_BACKEND=<backend_name>

Next, import the `backend` module using:
import geomstats.backend as gs
-->


## Description

The `Tree` class uses splits and edge lengths to represent unrooted phylogenetic trees without pendant edges.

Following Semple and Steel 2003 and [Lueg, 2023](https://ediss.uni-goettingen.de/handle/11858/14443), an unrooted phylogenetic tree is a tuple of edge splits and edge lengths; $T = (E, l)$ where $E \subset e\_compatible$ and $l \in (0, \infty)^{|E|}$.

Such a tree has N labels, $L_0=\{0, 1, ..., N-1\}$. Those N labels yield a set of possible edges, $\ e = \{A\|B : L_0 = A \bigcup B, \|A\|, \|B\| ≥ 2\}$. Note that this definition excludes pendant edges (those touching a leaf), yielding $\|e\| = 2^{N-1} - N - 3$ possible interior edges of which a maximum of $N-3$ are compatible at once.

BHV space is defined for $N\ge 4$, as any less will not yield any interior edges and thus be boringly Euclidean.

People working with phylogenetic trees might have additional requirements, and the way the code is now set up allows these to be added on top of the existing class structure, rather than needing to set up weird internal modifications. The possible additional constraints concern pendant edges and roots.

Pendant edges: No pendant edges is now the default, because often it is the distance between splits that matters in phylogenetic trees, not the distance between the split and the leaf, because the leaf is at an arbitrary time, the present. Too add lengths to the pendant edges we will create a space that is the Cartesian product of BHV space and Euclidean space of dimension $N$. Geodesics in this new space are just geodesics in pendant-less BHV space, with extra straight line segments in $(0, \infty)^{N}$ added on. This will be done in a subsequent PR. (add issue and tag?)

Roots: Unrooted is the default. There is a bijection between unrooted phylogenetic trees with $N$ labels and rooted ones with $N-1$ labels (just add a leaf, label it 0, and call it a day). We can add more explicit support for this during class construction in a subsequent PR. (add issue and tag?)



## Issue

Solves the following issues:
- TreeTopology allowed empty splits.
- Tree allowed non-positive lengths.
- Impossible to create star tree at origin -- Closes #2070  
- The generate_random_tree function produced trees that are not phylogenetic trees. It also allowed for trees with an empty set on one side of the split. That split, while a legal split, does not make sense in the context of edge length and should not exist.

## Additional context

TODO:
- I dislike that there is a separate trees.py file. Feels v unnecessary.
- I dislike generate_splits/delete_splits being separate. Maybe should be part of ForestTopology?
- Unsure how to write tests for this. Should add one verifying that only_internal_edges=True and only_internal_edges=False both work, but the current test for bhv_space and tree generation is not clear.
- Would like to do Cartesian product to handle pendant edges. This should be another PR.
- Should random_point be able to generate the star ever?? The probability of that is technically zero? Or is it? (Seems different if I think in terms of topologies or in terms of spaces)

abertics committed:
Modify generate_random_tree to produce legal phylogenetic trees

abertics committed:
Fix failed TreeSpace test due to improper  passing of downstream arguments

abertics committed:
Make dictionary creation fewer lines (not related)

abertics committed:
Remove print statements

luisfpereira commented on geomstats/geometry/stratified/bhv_space.py:
```diff
@@ -56,6 +62,13 @@ def generate_random_tree(n_labels, p_keep=0.9, btol=1e-8):
     initial_splits = generate_splits(labels)
     splits = delete_splits(initial_splits, labels, p_keep, check=False)
 
+    if only_internal_edges:
```
I think this should go before l63.

Better yet, would be to have it within `generate_splits`, as it would also solve the bug for `wald`.

A last question is: do we want to provide a boolean option or should we assume vertices of degree 1 are forbidden?

abertics replied:
I agree it should go within generate_splits. But I would prefer to add an argument like excluded_degrees, that is a list of positive integers. Then in the logic of creation we exclude the possibility of generating splits resulting in degrees we dislike. The problem is it is doable but non-trivial to find degree of node from the splits list. In practice, we will probably only care about excluding degree 1 and degree 2 nodes (xtree). 

This whole tree debacle should be an issue so that we can address it in more depth.

luisfpereira reviewed (changes requested):
Almost there @abertics.

Just a couple of comments to address (sorry the pedantism).

abertics commented on geomstats/geometry/stratified/bhv_space.py:
```diff
@@ -56,6 +62,13 @@ def generate_random_tree(n_labels, p_keep=0.9, btol=1e-8):
     splits = delete_splits(initial_splits, labels, p_keep, check=False)
 
+    if only_internal_edges:
+        splits[:] = [
```
This is a weird Python-specific operation called slice assignment, used to modify the contents of a list in-place, also changing the original object. I'll remove it, because I'm not sure it matters.

abertics committed:
Move singleton split checking to trees; ensure no empty splits

abertics committed:
Change nomenclature from interior to pendant edge

abertics committed:
Modify comment about pendant edges to be more clear

abertics committed:
Change name from interior to pendant in rancom_point function as well

abertics committed:
Rename from singleton to pendant

abertics committed:
Add outlier handling to Tree-related classes, including start tree creation and error handling. Make sanity tests in notebook

abertics committed:
Fix geodesic at origin bug by adding logic to generate star tree

abertics committed:
Add fake tests for metric dist and geodesic

abertics committed:
Add test at not beginning/middle/end

abertics committed:
Change some methods to static when able

abertics committed:
Modify generate_splits to exclude singletons; add tests

abertics committed:
Small syntax changes to make nicer

abertics committed:
Remove unnecessary list from tuple comprehension

abertics committed:
Rename example tree notebook

abertics committed:
Capitalise letter

luisfpereira commented on geomstats/geometry/stratified/bhv_space.py:
```diff
@@ -94,11 +72,33 @@ class TreeTopology(ForestTopology):
-            split_sets=(splits,),
-        )
+    def __init__(self, splits, n_labels=None):
+        if not splits:
```
undefined behavior if `n_labels=None`.

abertics replied:
do you want me to assert it is not None?

abertics replied:
I'll try something and see if it is satisfactory

luisfpereira commented on geomstats/geometry/stratified/bhv_space.py:
```diff
@@ -94,11 +72,33 @@ class TreeTopology(ForestTopology):
+                split_sets=(splits,),
+            )
+        else:
+            TreeTopology._check_no_pendant_edges(splits)
```
why not `self`?

luisfpereira replied:
this applies in other parts of the code too

my recommendation is to use `self`, even if it is a `staticmethod`

abertics replied:
i did it static bc i was copying 

    def __init__(self, splits, lengths, n_labels=None):
        Tree._check_valid_lengths(lengths, len(splits))

abertics replied:
I think it makes more sense to leave as Tree. or Tksladfj. rather than self, so its clear it isn't using an instance data. id say self makes sense if we want subclasses to be able to override, but in this case it's just helping us privately so meh.

luisfpereira commented on geomstats/geometry/stratified/bhv_space.py:
```diff
@@ -94,11 +72,33 @@ class TreeTopology(ForestTopology):
+                partition=(set(range(n_labels)),),
+                split_sets=(splits,),
+            )
+        else:
```
should check `n_labels` and splits compatibility if `n_labels is not None`?

luisfpereira reviewed (changes requested):
Thanks @abertics. The docstrings are really good!

Just an initially set of comments to fix something I've found while translating the notebook tests to pytest tests.

luisfpereira commented:
Also, can we add the beautiful description of this PR to the docstrings of the appropriate classes?

luisfpereira committed:
Translate Abbys notebook into pytest tests

luisfpereira commented:
Just translated the notebook into tests.

There's something failing sometimes. Run `pytest tests/tests_geomstats/test_geometry/test_stratified/test_bhv_space.py::TestBHVMetric`.

After we identify the issue and fix it, it is mostly ready to merge (great if we can update the docstrings as mentioned above before).

abertics committed:
Add logic for when n_labels is None, or when n_labels is not None and splits exists

luisfpereira committed:
Fix tree random generation in TreeSpace

abertics committed:
Update doc string verbosely

abertics commented:
I have added n_labels None logic. I've added stuff from PR description to doc string.

I have not changed static method call. I have also not fixed the test, because I can't seem to run the test with it failing.

abertics committed:
Disallow creation of tree with <4 labels

luisfpereira committed:
Cleanup before merge
